colbert/index_updater.py

```diff
@@ -0,0 +1,433 @@
+        
+        print(f"writing {pid_end - pid_start} passages to the last chunk...")
+        num_chunks = self.metadata['num_chunks']
+        
+        # append to current last chunk
+        curr_embs = ResidualEmbeddings.load(self.index_path, num_chunks - 1)
+        curr_embs.codes = torch.cat((curr_embs.codes, self.searcher.ranker.embeddings.codes[emb_start:emb_end]))
+        curr_embs.residuals = torch.cat((curr_embs.residuals, self.searcher.ranker.embeddings.residuals[emb_start:emb_end]))
+        path_prefix = os.path.join(self.index_path, f'{num_chunks - 1}')
+        curr_embs.save(path_prefix)
+        
+        # update doclen of last chunk
+        curr_doclens = self._load_chunk_doclens(num_chunks - 1).tolist()
+        curr_doclens.extend(self.searcher.ranker.doclens.tolist()[pid_start:pid_end])
+        doclens_path = os.path.join(self.index_path, f'doclens.{num_chunks - 1}.json')
+        with open(doclens_path, 'w') as output_doclens:
+            ujson.dump(curr_doclens, output_doclens)
+        
+        # update metadata of last chunk
+        print("updating metadata of current last chunk...")
+        chunk_metadata = self._load_chunk_metadata(num_chunks - 1)
+        chunk_metadata['num_passages'] += pid_end - pid_start
+        chunk_metadata['num_embeddings'] += emb_end - emb_start
+        chunk_metadata_path = os.path.join(self.index_path, f'{num_chunks - 1}.metadata.json')
+        with open(chunk_metadata_path, 'w') as output_chunk_metadata:
+            ujson.dump(chunk_metadata, output_chunk_metadata)
+            
+    def _write_to_new_chunk(self, chunk_idx, pid_start, pid_end, emb_start, emb_end):
+        # Helper function for IndexUpdater.persist_to_disk()
+        
+        # save embeddings to new chunk
+        print(f'writing {pid_end - pid_start} passages to chunk {chunk_idx}...')
+        curr_embs = ResidualEmbeddings(self.searcher.ranker.embeddings.codes[emb_start:emb_end], self.searcher.ranker.embeddings.residuals[emb_start:emb_end])
+        path_prefix = os.path.join(self.index_path, f'{chunk_idx}')
+        curr_embs.save(path_prefix)
+        
+        # create doclen json file for new chunk
+        curr_doclens = self.searcher.ranker.doclens.tolist()[pid_start:pid_end]
+        doclens_path = os.path.join(self.index_path, f'doclens.{chunk_idx}.json')
+        with open(doclens_path, 'w+') as output_doclens:
+            ujson.dump(curr_doclens, output_doclens)
+            
+        # create metadata json file for new chunk
+        chunk_metadata = {'passage_offset': pid_start, 'num_passages': pid_end - pid_start, 'embedding_offset':emb_start , 'num_embeddings': emb_end - emb_start}
+        chunk_metadata_path = os.path.join(self.index_path, f'{chunk_idx}.metadata.json')
+        with open(chunk_metadata_path, 'w+') as output_chunk_metadata:
+            ujson.dump(chunk_metadata, output_chunk_metadata)
+        
+    def _remove_passage_from_disk(self, pid):
```

Comment:
Loading and dumping large files once per pid is not scalable.

That said, I'm happy with the current single-pid implementation (for loop) as a (permanent) "beta release" (i.e., use at your own risk; no guarantees it's fast at scale, but it works).

Noting here: If you'd like this to be efficient, you'll need to basically gather all pids that have the same chunk_idx and then do the rest of the steps in a batch.

Reply:
cc: @santhnm2